src/Gaufrette/Adapter/Ftp.php

```diff
@@ -54,93 +55,122 @@ public function __construct($directory, $host, $options = array())
     }
 
     /**
-     * {@inheritDoc}
+     * @param string $key
+     * @return bool|string
      */
     public function read($key)
     {
-        $this->ensureDirectoryExists($this->directory, $this->create);
+        //change to the directory that contain the file
+        $this->moveToTargetDirectory($key);
+        $file = basename($key);
 
         $temp = fopen('php://temp', 'r+');
 
-        if (!ftp_fget($this->getConnection(), $temp, $this->computePath($key), $this->mode)) {
+        if (!ftp_fget($this->getConnection(), $temp, $file, $this->mode)) {
+            //change back to ftp root directory
+            $this->changeDirectory($this->directory);
             return false;
         }
 
         rewind($temp);
         $contents = stream_get_contents($temp);
+        //change back to ftp root directory
+        $this->changeDirectory($this->directory);
```

Comment:
I would not put this in the middle of the code retrieving the content from ``$temp``, for readability